src/main/java/com/google/research/bleth/services/DatabaseService.java

```diff
@@ -0,0 +1,157 @@
+     * @param round is the round number in the simulation that matches the input board state.
+     * @param board is a Board object representing the state of the board.
+     * @throws PreparedQuery.TooManyResultsException if a board state with the same simulationId and round already exists.
+     */
+    public void writeRealBoardState(String simulationId, int round, Board board)
+            throws PreparedQuery.TooManyResultsException {
+        writeBoardState(simulationId, round, board, "TracingRealBoardState");
+    }
+
+    /**
+     * Create an entity of a estimated board state based on input params and write it to datastore.
+     * @param simulationId is the Id of the simulation associated with the input board state.
+     * @param round is the round number in the simulation that matches the input board state.
+     * @param board is a Board object representing the state of the board.
+     * @throws PreparedQuery.TooManyResultsException if a board state with the same simulationId and round already exists.
+     */
+    public void writeEstimatedBoardState(String simulationId, int round, Board board)
+            throws PreparedQuery.TooManyResultsException {
+        writeBoardState(simulationId, round, board, "TracingEstimatedBoardState");
+    }
+
+    private String getBoardState(String simulationId, int round, String queryKind)
+            throws PreparedQuery.TooManyResultsException {
+
+        // Set simple predicates.
+        Query.FilterPredicate filterBySimulationId =
+                new Query.FilterPredicate("simulationId", Query.FilterOperator.EQUAL, simulationId);
+        Query.FilterPredicate filterByRound =
+                new Query.FilterPredicate("round", Query.FilterOperator.EQUAL, round);
+
+        // Compose simple predicates.
+        Query.CompositeFilter composedFilter = Query.CompositeFilterOperator.and(filterBySimulationId, filterByRound);
+
+        // Create query and return result.
+        Query boardStateBySimulationAndRoundQuery = new Query(queryKind).setFilter(composedFilter);
+        PreparedQuery pq = datastore.prepare(boardStateBySimulationAndRoundQuery);
+
+        Entity boardStateEntity = pq.asSingleEntity();
+        if (boardStateEntity == null) {
+            return null;
+        } else {
+            return (String) boardStateEntity.getProperty("state");
+        }
+    }
+
+    /**
+     * Retrieve the real board state who matches the input params (if exists) and return it as a JSON string.
+     * @param simulationId is the Id of the simulation.
+     * @param round is the round number in the simulation.
+     * @return a JSON string representing the corresponding board state, or null if it doesn't exist.
+     * @throws PreparedQuery.TooManyResultsException if there are more than a single entity matching the input params.
+     */
+    public String getRealBoardState(String simulationId, int round)
```

Comment:
The `write` method gets a board and write to the db a JSON string.
The `read` method gets a simulation ID and round number and retrieves the JSON string representing this board state.

The reason `read` returns a JSON string and not a Board object is that the string can be written to a `HTTPServletResponse` and it is the format the UI code expects to receive in order to visualize the board state:

```java
// A servlet for reading a board state.
String boardState = DatabaseService.getInstance().readRealBoardState(simulationId, round);
response.getWriter().println(boardState);
```

If we change the `read` method return value to be a Board, we will have to add another phase of translating from Board to JSON string outside the `read` method:

```java
// A servlet for reading a board state.
Board boardStateAsBoard= DatabaseService.getInstance().readRealBoardState(simulationId, round);
String boardStateAsBoardJson = BoardParser.encode(boardStateAsBoard) // <---
response.getWriter().println(boardStateAsBoardJson);
```
I personally prefer the first implementation (unbalanced, but requires less operations inside the servlet), please let me know if there are any other advantages of the second one.

Reply:
There are two issues with this approach:
1. It hardcodes the assumption that the database format is one-and-the-same as the communication format. If we ever decide to vary them, even slightly - this entire structure breaks. (Google best practices call for e.g. using different protos at these layers - precisely because changes do happen, sometimes, and if you're not ready for them the change becomes extra-costly.)
2. If we ever implement additional logic in the server that reads from the database (e.g. "continue a saved simulation for another 1000 steps") - the JSON-returning read() makes this harder than it should be.

Reply:
Do we really want to return a `Board` without an associated `Simulation` object? 

If we do: this means that when creating the returned `Board` we are creating `Agent` objects with default movement and awakeness strategies (e.g stationary), and a `null` value where should be a reference to the `Simulation` object.

If we don't: that means we need to store in the db all information required for re-generating the `Simulation` object, including:
- Agent's movement stratregies.
- Observers' awakeness strategies.
- Resolver current mapping state (which is not only the estimated board state, but also the transmission received in in the last round).

In such case, the `DatabaseService` class will have to be updated whenever a new simulation class is being pushed to the master, and also a [JsonSeriallizer](https://www.javadoc.io/doc/com.google.code.gson/gson/2.8.5/com/google/gson/JsonSerializer.html) and a [JsonDeserializer](https://www.javadoc.io/doc/com.google.code.gson/gson/2.8.5/com/google/gson/JsonDeserializer.html) should be created for each class we want to be able to write to the db.

Most of this additional data to be stored is not relevant for visualization, but only for creating the returned board. Moreover, if we return a board associated to a simulation, when and where this simulation is created?